transcom/mymove
# MB-7933 Create segment for SAC/SDN when it is present

## [Jira ticket](https://dp3.atlassian.net/browse/MB-7933)

## Summary

This adds the SAC/SDN to EDI for each shipment service item in an additional FA2 segment with the ZZ qualifier `FA2*ZZ*GOESHERE`. (SAC and SDN live in the same field in MM)

If SAC/SDN is nil or an empty string, that segment should be omitted, if over 80 chars, it should show an error message.

Depending on` SACtype`, you'll see either the HHG (default) or NTS SAC/SDN.

## Verification Steps for the Author

These are to be checked by the author.


## Verification Steps for Reviewers

These are to be checked by a reviewer.

### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/run-storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/about/application-setup/milmove-local-client/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/about/development/testing)

### How to test

1. Go to the office app and Login as a TIO with gbloc KKFA
2. Search for move code "RDY4PY" and click in
3. Scroll down to payment requests, click Review, and approve 1+.
4. Refer to the [Acceptance Testing doc](https://transcom.github.io/mymove-docs/docs/backend/testing/acceptance-testing-syncada-edi-invoicing). 
5. In your terminal, if you have the json file from the doc in root dir, run `prime-api-client --insecure create-payment-request --filename ./payload.json | jq '.id,.paymentRequestNumber'` which should give you the 9 digit payment request ID for the following command
6. Run `make bin/prime-api-client bin/generate-payment-request-edi && generate-payment-request-edi --payment-request-number 6682-5285-1` replacing the end numbers with the payment request ID from previous
7. Confirm that the EDI prints in your terminal and does NOT include the new FA2 field, as hhg SAC's empty.
8. Go back into the move and add HHG and NTS SACs
9. Run 5 again and check that HHG SAC's in a new FA2 field

<img width="247" alt="Screenshot 2023-05-18 at 12 17 00 PM" src="https://github.com/transcom/mymove/assets/108361430/68b4752c-766e-421d-9537-8f65e411ddca">

10. Make the hhg SACs 81+ chars and test 5 again. you should get an error msg in console noting the value's over 80 chars.
<img width="536" alt="Screenshot 2023-05-19 at 9 06 10 AM" src="https://github.com/transcom/mymove/assets/108361430/c665e4ff-37a7-45bd-abfe-36c62cbc0ed9">

11. Go to the DB and change SACtype to NTS (see screenshot)

<img width="457" alt="Screenshot 2023-05-19 at 9 20 54 AM" src="https://github.com/transcom/mymove/assets/108361430/854e1e4d-bb35-4086-9a5d-3ce1d09adc6b">

12. Run 5 again and check that NTS SAC, not HHG one, is in a new FA2 field
13. Make the nts SAC 81+ chars and test 5 again. you should get an error msg in console noting the value's over 80 chars.



### Frontend


### Backend


### Database

#### Any new migrations/schema changes:


## Screenshots
